src/wasm-type.h

```diff
@@ -164,12 +192,147 @@ struct Signature {
+  std::string toString() const;
+};
+
+struct Field {
+  Type type;
+  bool mutable_;
+  Field(Type type, bool mutable_ = false) : type(type), mutable_(mutable_) {}
+  bool operator==(const Field& other) const {
+    return type == other.type && mutable_ == other.mutable_;
+  }
+  bool operator!=(const Field& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+typedef std::vector<Field> FieldList;
+
+struct Struct {
+  FieldList fields;
+  bool nullable;
+  Struct(const Struct& other)
+    : fields(other.fields), nullable(other.nullable) {}
+  Struct(FieldList fields, bool nullable = false)
+    : fields(fields), nullable(nullable) {}
+  bool operator==(const Struct& other) const {
+    return fields == other.fields && nullable == other.nullable;
+  }
+  bool operator!=(const Struct& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+struct Array {
+  Field element;
+  bool nullable;
+  Array(const Array& other)
+    : element(other.element), nullable(other.nullable) {}
+  Array(Field element, bool nullable = false)
+    : element(element), nullable(nullable) {}
+  bool operator==(const Array& other) const {
+    return element == other.element && nullable == other.nullable;
+  }
+  bool operator!=(const Array& other) const { return !(*this == other); }
+  std::string toString() const;
+};
+
+union TypeDef {
+  enum Kind { TupleKind, SignatureKind, StructKind, ArrayKind };
+
+  struct TupleDef {
+    Kind kind;
+    Tuple tuple;
+  } tupleDef;
+  struct SignatureDef {
+    Kind kind;
+    Signature signature;
+  } signatureDef;
+  struct StructDef {
+    Kind kind;
+    Struct struct_;
+  } structDef;
+  struct ArrayDef {
+    Kind kind;
+    Array array;
+  } arrayDef;
+
+  TypeDef(Tuple tuple) : tupleDef{TupleKind, tuple} {}
+  TypeDef(Signature signature) : signatureDef{SignatureKind, signature} {}
+  TypeDef(Struct struct_) : structDef{StructKind, struct_} {}
+  TypeDef(Array array) : arrayDef{ArrayKind, array} {}
+  TypeDef(const TypeDef& other) {
+    switch (other.getKind()) {
+      case TupleKind:
+        ::new (&tupleDef) auto(other.tupleDef);
+        break;
+      case SignatureKind:
+        ::new (&signatureDef) auto(other.signatureDef);
+        break;
+      case StructKind:
+        ::new (&structDef) auto(other.structDef);
+        break;
+      case ArrayKind:
+        ::new (&arrayDef) auto(other.arrayDef);
+        break;
+      default:
+        WASM_UNREACHABLE("unexpected kind");
+    }
+  }
+  ~TypeDef() {
+    switch (getKind()) {
+      case TupleKind: {
+        tupleDef.~TupleDef();
+        break;
+      }
+      case SignatureKind: {
+        signatureDef.~SignatureDef();
+        break;
+      }
+      case StructKind: {
+        structDef.~StructDef();
+        break;
+      }
+      case ArrayKind: {
+        arrayDef.~ArrayDef();
+        break;
+      }
+      default:
+        WASM_UNREACHABLE("unexpected kind");
+    }
+  }
```

Comment:
This is unfortunate noise, but I guess it's necessary while we don't have access to C++17's `std::variant`?